pyccel/ast/builtins.py

```diff
@@ -715,6 +716,32 @@ def is_homogeneous(self):
         """
         return True
 
+
+class PythonListFunction(TypedAstNode):
+    """
+    Class representing a call to the `list` function.
+
+    Class representing a call to the `list` function. This is
+    different to the `[,]' syntax as it only takes one argument
+    and unpacks any variables.
+
+    Parameters
+    ----------
+    arg : TypedAstNode
+        The argument passed to the function call.
+    """
+    __slots__ = ()
+    _attribute_nodes = ()
+
+    def __new__(cls, arg):
+        if isinstance(arg, PythonList):
+            return arg
+        elif isinstance(arg.shape[0], LiteralInteger):
+            return PythonList(*[arg[i] for i in range(arg.shape[0])])
+        else:
+            raise TypeError(f"Can't unpack {arg} into a list")
```

Comment:
As implemented this function will only handle arguments of known size. For example the following code won't translate:
```python
def f(a : list[int]):
    return list(a)
```
it raises:
```
ERROR at annotation (semantic) stage
pyccel:
 |fatal [semantic]: tmp_list_copy.py [1,10]| Cannot unpack a into a list
Pyccel has encountered syntax that has not been implemented yet. Please create an issue at https://github.com/pyccel/pyccel/issues and provide a small example of your problem. Do not forget to specify your target language. (list[int])
```

The fix should be similar to what is done in #1897 .

Reply:
@EmilyBourne , I am trying to fix the issue. After some changes I am getting another exception which is as follows:

 |fatal [semantic]: test.py [2,10]| Cannot handle non-homogenous type index
Pyccel has encountered syntax that has not been implemented yet. Please create an issue at https://github.com/pyccel/pyccel/issues and provide a small example of your problem. Do not forget to specify your target language. (list[int])

As I understand the lists should be homogeneous. Should this PR handle this also?

Reply:
I've had a quick look and it seems the error is coming from [here](https://github.com/pyccel/pyccel/blob/5a75f34563bd1ea3287afc29949bca98400be6fe/pyccel/parser/semantic.py#L1940). This function is designed to handle type annotations. Type annotations for lists are only added in #1847.

Sorry, I should have given you a better test case. Here is a new one:
```python
def f(a : 'int[:]'):
    return list(a)
```
where `a` is a NumPy array. The type annotations are properly implemented here

Reply:
@shoaib-moeen if you merge the main branch the error you were seeing should now be fixed.